src/ipc/ipc4/helper.c

```diff
@@ -571,6 +566,59 @@ int ipc4_chain_dma_state(struct comp_dev *dev, struct ipc4_chain_dma *cdma)
 }
 #endif
 
+static int ipc4_update_comps_direction(struct ipc *ipc, uint32_t ppl_id)
+{
+	struct ipc_comp_dev *icd;
+	struct list_item *clist;
+	struct comp_buffer *src_buf;
+
+	list_for_item(clist, &ipc->comp_list) {
+		icd = container_of(clist, struct ipc_comp_dev, list);
+		if (icd->type != COMP_TYPE_COMPONENT)
+			continue;
+
+		if (dev_comp_pipe_id(icd->cd) != ppl_id)
+			continue;
+
+		if (icd->cd->direction_set)
+			continue;
+
+		src_buf = list_first_item(&icd->cd->bsource_list, struct comp_buffer, sink_list);
```

Comment:
cannot you hit a component with an empty source list?

Reply:
exactly, that's what I mean: a freshly initialised list with no elements added to it is empty. For such a list `list_first_item()` would return an invalid pointer - a pointer to an imaginary object of the requested type "around" the list head itself.